polynote-spark-runtime/src/main/scala/polynote/runtime/spark/reprs/SparkReprsOf.scala

```diff
@@ -3,74 +3,181 @@ package polynote.runtime.spark.reprs
+          case sparkTypes.BinaryType  => BinaryType -> ((out, row) => DataEncoder.byteArray.encode(out, row.getBinary(index)))
+          case sparkTypes.StringType  => StringType -> ((out, row) => DataEncoder.string.encode(out, row.getString(index)))
+          case struct @ sparkTypes.StructType(_) =>
+            val (structType, encode) = dataTypeAndEncoder(struct)
+            structType -> ((out, row) => encode(out, row.getStruct(index, struct.fields.length)))
+        }.map {
+          case (dt, enc) => StructField(name, dt) -> enc
+        }
+    }.unzip
 
-  def instance[T](reprs: T => Array[ValueRepr]): SparkReprsOf[T] = new SparkReprsOf[T] {
-    def apply(value: T): Array[ValueRepr] = reprs(value)
+    StructType(fieldTypes.toList) -> {
+      (out, row) =>
+        fieldEncoders.foreach(_.apply(out, row))
+    }
   }
 
-  implicit val dataFrame: SparkReprsOf[DataFrame] = {
-    import org.apache.spark.sql.{types => sparkTypes}
-    import org.apache.spark.sql.catalyst.InternalRow
-
-    def dataTypeAndEncoder(schema: sparkTypes.StructType): (StructType, (DataOutput, InternalRow) => Unit) = {
-      val (fieldTypes, fieldEncoders) = schema.fields.zipWithIndex.flatMap {
-        case (sparkTypes.StructField(name, dataType, nullable, _), index) =>
-          Option(dataType).collect[(DataType, (DataOutput, InternalRow) => Unit)] {
-            case sparkTypes.ByteType    => ByteType -> ((out, row) => DataEncoder.byte.encode(out, row.getByte(index)))
-            case sparkTypes.BooleanType => BoolType -> ((out, row) => DataEncoder.boolean.encode(out, row.getBoolean(index)))
-            case sparkTypes.ShortType   => ShortType -> ((out, row) => DataEncoder.short.encode(out, row.getShort(index)))
-            case sparkTypes.IntegerType => IntType -> ((out, row) => DataEncoder.int.encode(out, row.getInt(index)))
-            case sparkTypes.LongType    => LongType ->((out, row) => DataEncoder.long.encode(out, row.getLong(index)))
-            case sparkTypes.FloatType   => FloatType -> ((out, row) => DataEncoder.float.encode(out, row.getFloat(index)))
-            case sparkTypes.DoubleType  => DoubleType -> ((out, row) => DataEncoder.double.encode(out, row.getDouble(index)))
-            case sparkTypes.BinaryType  => BinaryType -> ((out, row) => DataEncoder.byteArray.encode(out, row.getBinary(index)))
-            case sparkTypes.StringType  => StringType -> ((out, row) => DataEncoder.string.encode(out, row.getString(index)))
-            case struct @ sparkTypes.StructType(_) =>
-              val (structType, encode) = dataTypeAndEncoder(struct)
-              structType -> ((out, row) => encode(out, row.getStruct(index, struct.fields.length)))
-          }.map {
-            case (dt, enc) => StructField(name, dt) -> enc
-          }
-      }.unzip
+  private[polynote] class FixedSizeDataFrameDecoder(structType: StructType, encode: (DataOutput, InternalRow) => Unit) extends (InternalRow => Array[Byte]) with Serializable {
```

Comment:
is this just more efficient than having everything be a `VariableSizeDataFrameDecoder`?

Reply:
Well, the fixed size one can allocate ahead-of-time. This one has to use `ByteArrayOutputStream` which re-allocates as it grows.

Reply:
BTW, the reason these are made their own classes is to prevent closure serializability issues.